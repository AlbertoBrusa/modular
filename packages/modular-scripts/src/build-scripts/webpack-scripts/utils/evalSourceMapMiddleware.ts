import Server from 'webpack-dev-server';
import type {
  NextFunction,
  Request,
  Response,
  Stats,
} from 'webpack-dev-server';
import webpack from 'webpack';

// import type Server from 'webpack-dev-server';
// / <reference path="node_modules/webpack/types.d.ts" />
function base64SourceMap(source: webpack.sources.Source) {
  const base64 = Buffer.from(JSON.stringify(source.map()), 'utf8').toString(
    'base64',
  );
  return `data:application/json;charset=utf-8;base64,${base64}`;
}

function getSourceById(stats: Stats, id: string) {
  const module = Array.from(stats.compilation.modules).find(
    (m) => stats.compilation.chunkGraph.getModuleId(m) === id,
  );
  return module ? module.originalSource() : null;
}

/*
 * Middleware responsible for retrieving a generated source
 * Receives a webpack internal url: "webpack-internal:///<module-id>"
 * Returns a generated source: "<source-text><sourceMappingURL><sourceURL>"
 *
 * Based on EvalSourceMapDevToolModuleTemplatePlugin.js
 */
export default function createEvalSourceMapMiddleware(server: Server) {
  return function handleWebpackInternalMiddleware(
    req: Request,
    res: Response,
    next: NextFunction,
  ) {
    if (req.url.startsWith('/__get-internal-source')) {
      const fileName = req.query.fileName as string;
      const match = fileName.match(/webpack-internal:\/\/\/(.+)/);
      const id = match ? match[1] : undefined;
      const stats = server['stats'] as Stats | undefined;
      if (!id || !stats) {
        next();
      } else {
        const source = getSourceById(stats, id);
        if (source) {
          const sourceMapURL = `//# sourceMappingURL=${base64SourceMap(
            source,
          )}`;
          const sourceURL = `//# sourceURL=webpack-internal:///${module.id}`;
          res.end(
            `${source.source() as string}\n${sourceMapURL}\n${sourceURL}`,
          );
        } else {
          next();
        }
      }
    } else {
      next();
    }
  };
}
