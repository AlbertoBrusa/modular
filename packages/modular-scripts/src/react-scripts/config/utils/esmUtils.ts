const packageRegex =
  /^(@[a-z0-9-~][a-z0-9-._~]*)?\/?([a-z0-9-~][a-z0-9-._~]*)\/?(.*)/;

export function parsePackageName(name: string) {
  const parsedName = packageRegex.exec(name);
  if (!parsedName) {
    return;
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [_, scope, module, submodule] = parsedName;
  const dependencyName = (scope ? `${scope}/` : '') + module;
  return { dependencyName, scope, module, submodule };
}
