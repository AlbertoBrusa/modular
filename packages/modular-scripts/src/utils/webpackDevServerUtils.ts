// Ported over functions from webpackDevServerUtils.ts
import address from 'address';
import chalk from 'chalk';
import url from 'url';
import path from 'path';
import { existsSync } from 'fs';
import { log } from './logger';
import { ProxyConfigArray, Request, Response } from 'webpack-dev-server';
import type * as http from 'http';
import type * as httpProxy from 'http-proxy';

export function prepareProxy(
  proxy: unknown,
  appPublicFolder: string,
  servedPathname: string,
): ProxyConfigArray | undefined {
  // `proxy` lets you specify alternate servers for specific requests.
  if (!proxy) {
    return undefined;
  }
  if (typeof proxy !== 'string') {
    log(chalk.red('When specified, "proxy" in package.json must be a string.'));
    log(chalk.red('Instead, the type of "proxy" was "' + typeof proxy + '".'));
    log(
      chalk.red(
        'Either remove "proxy" from package.json, or make it a string.',
      ),
    );
    process.exit(1);
  }

  // If proxy is specified, let it handle any request except for
  // files in the public folder and requests to the WebpackDevServer socket endpoint.
  // https://github.com/facebook/create-react-app/issues/6720
  const sockPath = process.env.WDS_SOCKET_PATH || '/ws';
  const isDefaultSockHost = !process.env.WDS_SOCKET_HOST;
  function mayProxy(pathname: string) {
    const maybePublicPath = path.resolve(
      appPublicFolder,
      pathname.replace(new RegExp('^' + servedPathname), ''),
    );
    const isPublicFileRequest = existsSync(maybePublicPath);
    // used by webpackHotDevClient
    const isWdsEndpointRequest =
      isDefaultSockHost && pathname.startsWith(sockPath);
    return !(isPublicFileRequest || isWdsEndpointRequest);
  }

  if (!/^http(s)?:\/\//.test(proxy)) {
    console.log(
      chalk.red(
        'When "proxy" is specified in package.json it must start with either http:// or https://',
      ),
    );
    process.exit(1);
  }

  const target = process.platform === 'win32' ? resolveLoopback(proxy) : proxy;

  return [
    {
      target,
      logLevel: 'silent',
      // For single page apps, we generally want to fallback to /index.html.
      // However we also want to respect `proxy` for API calls.
      // So if `proxy` is specified as a string, we need to decide which fallback to use.
      // We use a heuristic: We want to proxy all the requests that are not meant
      // for static assets and as all the requests for static assets will be using
      // `GET` method, we can proxy all non-`GET` requests.
      // For `GET` requests, if request `accept`s text/html, we pick /index.html.
      // Modern browsers include text/html into `accept` header when navigating.
      // However API calls like `fetch()` won’t generally accept text/html.
      // If this heuristic doesn’t work well for you, use `src/setupProxy.js`.
      context: function (pathname: string, req: Request): boolean {
        const shouldFilter =
          req.method !== 'GET' ||
          (mayProxy(pathname) &&
            req.headers.accept &&
            req.headers.accept.indexOf('text/html') === -1)
            ? true
            : false;
        return shouldFilter;
      },
      onProxyReq: function (
        proxyReq: http.ClientRequest,
        _req: Request,
        _res: Response,
        _options: httpProxy.ServerOptions,
      ) {
        // Browsers may send Origin headers even with same-origin
        // requests. To prevent CORS issues, we have to change
        // the Origin to match the target URL.
        if (proxyReq.getHeader('origin')) {
          proxyReq.setHeader('origin', target);
        }
      },
      onError: onProxyError(target),
      secure: false,
      changeOrigin: true,
      ws: true,
      xfwd: true,
    },
  ];
}

function resolveLoopback(proxy: string) {
  const o = url.parse(proxy);
  if (o.hostname !== 'localhost') {
    return proxy;
  }
  // Unfortunately, many languages (unlike node) do not yet support IPv6.
  // This means even though localhost resolves to ::1, the application
  // must fall back to IPv4 (on 127.0.0.1).
  // We can re-enable this in a few years.
  /*try {
      o.hostname = address.ipv6() ? '::1' : '127.0.0.1';
    } catch (_ignored) {
      o.hostname = '127.0.0.1';
    }*/

  try {
    // Check if we're on a network; if we are, chances are we can resolve
    // localhost. Otherwise, we can just be safe and assume localhost is
    // IPv4 for maximum compatibility.
    if (!address.ip()) {
      o.hostname = '127.0.0.1';
    }
  } catch (_ignored) {
    o.hostname = '127.0.0.1';
  }
  return url.format(o);
}

// We need to provide a custom onError function for httpProxyMiddleware.
// It allows us to log custom error messages on the console.
function onProxyError(proxy: string) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (err: any, req: Request, res: Response) => {
    const host = req.headers && req.headers.host;
    console.log(
      chalk.red('Proxy error:') +
        ' Could not proxy request ' +
        chalk.cyan(req.url) +
        ' from ' +
        chalk.cyan(host) +
        ' to ' +
        chalk.cyan(proxy) +
        '.',
    );
    console.log(
      'See https://nodejs.org/api/errors.html#errors_common_system_errors for more information (' +
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        chalk.cyan(err.code) +
        ').',
    );
    console.log();

    // And immediately send the proper error response to the client.
    // Otherwise, the request will eventually timeout with ERR_EMPTY_RESPONSE on the client side.
    if (res.writeHead && !res.headersSent) {
      res.writeHead(500);
    }
    res.end(
      // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
      'Proxy error: Could not proxy request ' +
        req.url +
        ' from ' +
        host +
        ' to ' +
        proxy +
        ' (' +
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        err.code +
        ').',
    );
  };
}
