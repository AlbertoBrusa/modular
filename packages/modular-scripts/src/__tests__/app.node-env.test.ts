/* eslint-disable jest/no-commented-out-tests */
import tree from 'tree-view-for-tests';
import path from 'path';
import fs from 'fs-extra';
import prettier from 'prettier';
import { createModularTestContext, runModularStreamlined } from '../test/utils';

// These tests must be executed sequentially with `--runInBand`.

// Temporary test context paths set by createModularTestContext()
let tempModularRepo: string;

describe('when working with a NODE_ENV app', () => {
  describe('WHEN building with webpack', () => {
    beforeAll(async () => {
      await setupNodeEnvApp();
      await runModularStreamlined(tempModularRepo, 'build node-env-app');
    });

    // it('can build a app', () => {
    //   expect(tree(path.join(tempModularRepo, 'dist', 'node-env-app')))
    //     .toMatchInlineSnapshot(`
    //     "node-env-app
    //     ├─ asset-manifest.json #1oz8cie
    //     ├─ favicon.ico #6pu3rg
    //     ├─ index.html #8yl21p
    //     ├─ logo192.png #1nez7vk
    //     ├─ logo512.png #1hwqvcc
    //     ├─ manifest.json #19gah8o
    //     ├─ package.json
    //     ├─ robots.txt #1sjb8b3
    //     └─ static
    //        └─ js
    //           ├─ main.5d879077.js #u1fxs5
    //           ├─ main.5d879077.js.map #158jj7c
    //           ├─ runtime-main.97707f9d.js #15lezt9
    //           └─ runtime-main.97707f9d.js.map #1yg8f1m"
    //   `);
    // }); c42f5e48

    it('can generate a js/main.5d879077.js.map', async () => {
      expect(
        String(
          await fs.readFile(
            path.join(
              tempModularRepo,
              'dist',
              'node-env-app',
              'static',
              'js',
              'main.5d879077.js.map',
            ),
          ),
        ),
      ).toMatchSnapshot();
    });
    it('can generate a js/main.5d879077.js', async () => {
      expect(
        String(
          await fs.readFile(
            path.join(
              tempModularRepo,
              'dist',
              'node-env-app',
              'static',
              'js',
              'main.5d879077.js',
            ),
          ),
        ),
      ).toMatchSnapshot();
    });
  });
  describe('WHEN building with esbuild', () => {
    beforeAll(async () => {
      await setupNodeEnvApp();
      await runModularStreamlined(tempModularRepo, 'build node-env-app', {
        env: {
          USE_MODULAR_ESBUILD: 'true',
        },
      });
    });

    it('can build a app', () => {
      expect(tree(path.join(tempModularRepo, 'dist', 'node-env-app')))
        .toMatchInlineSnapshot(`
        "node-env-app
        ├─ favicon.ico #6pu3rg
        ├─ index.html #177xfr4
        ├─ logo192.png #1nez7vk
        ├─ logo512.png #1hwqvcc
        ├─ manifest.json #19gah8o
        ├─ package.json
        ├─ robots.txt #1sjb8b3
        └─ static
           └─ js
              ├─ index-L2US6VEL.js #7mn428
              └─ index-L2US6VEL.js.map #sr6neg"
      `);
    });

    it('can generate a js/index-L2US6VEL.js', async () => {
      expect(
        prettier.format(
          String(
            await fs.readFile(
              path.join(
                tempModularRepo,
                'dist',
                'node-env-app',
                'static',
                'js',
                'index-L2US6VEL.js',
              ),
            ),
          ),
          {
            filepath: 'index-L2US6VEL.js',
          },
        ),
      ).toMatchSnapshot();
    });
  });
});

/**
 * Create temp modular repo and add an app called node-env-app
 */
async function setupNodeEnvApp(): Promise<void> {
  tempModularRepo = createModularTestContext();
  await runModularStreamlined(
    tempModularRepo,
    'add node-env-app --unstable-type app',
  );

  await fs.writeFile(
    path.join(tempModularRepo, 'packages', 'node-env-app', 'src', 'index.ts'),
    `
      console.log(process.env.NODE_ENV);
      export {};
      `,
  );
}
