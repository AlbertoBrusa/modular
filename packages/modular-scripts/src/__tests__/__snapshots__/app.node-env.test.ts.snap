// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when working with a NODE_ENV app WHEN building with esbuild can generate a js/index-L2US6VEL.js 1`] = `
"console.log("production");
//# sourceMappingURL=/static/js/index-L2US6VEL.js.map
"
`;
